### PR TITLE
[IMP] queue_job: detect jobs runned by workers that have been killed

### PR DESCRIPTION
**Pull request by AnizR**
## Goal
Do not keep jobs in 'started' status if their associated worker have been killed by a timeout and directly set it as 'failed'

### CODE REVIEW

File: queue_job/models/queue_job.py [hunk 1]
```diff
@@ -417,6 +419,37 @@ def autovacuum(self):
...
+                ("state", "=", "started"),
+                ("worker_pid", "!=", False),
+            ]
+        )
+
+        failed_jobs = self.env["queue.job"].browse()
+
+        for job in jobs:
+            if not psutil.pid_exists(job.worker_pid):
+                _logger.info("a process with pid %d does not exist" % job.worker_pid)
```
Comment (amh-mw) [inline queue_job/models/queue_job.py]: If I saw this in the logs, I don't think I'd know what to look for. Maybe say "a queue_job process" instead?

File: queue_job/models/queue_job.py [hunk 2]
```diff
@@ -417,6 +419,37 @@ def autovacuum(self):
...
+                failed_jobs += job
+        if failed_jobs:
+            for job in failed_jobs:
```
Comment (amh-mw) [inline queue_job/models/queue_job.py]: Why a second loop? I don't see any value in accumulating failed_jobs; just set the jobs failed as they are found?

**Review by amh-mw — APPROVED** [review]
This is a recurring problem for me, so I will backport to 15.0 shortly after merge.

File: queue_job/models/queue_job.py (hunk 1 above)
Comment (AnizR) [reply queue_job/models/queue_job.py]: Oops, i forgot to remove it, it to ensure that it was working during my tests :smile:

File: queue_job/models/queue_job.py (hunk 2 above)
Comment (AnizR) [reply queue_job/models/queue_job.py]: Indeed, not necessary, thanks!

File: queue_job/models/queue_job.py (hunk 1 above)
Comment (amh-mw) [reply queue_job/models/queue_job.py]: I do think there is value in logging unexpected queue_job workers deaths with greater specificity than the generic werkzeug worker timeout message.

Comment (AnizR) [reply queue_job/models/queue_job.py]: I see, I updated the message to give the worker pid and the job uuid.

**Review by guewen — COMMENTED** [review]
Hi, sadly the pid may be on another host and still executing

**Comment by AnizR** [comment]
> Hi, sadly the pid may be on another host and still executing

What would you advise ?
Making the server update the job before killing the worker? (extending the function[ `process_timeout` ](https://github.com/odoo/odoo/blob/16.0/odoo/service/server.py#L881)? I have no idea if it is possible )

**Comment by amh-mw** [comment]
> Hi, sadly the pid may be on another host and still executing

Record the hostname / other signature of the host and filter accordingly? Config parameter gate this feature?

**Comment by sbidoul** [comment]
@AnizR I think https://github.com/OCA/queue/pull/423 was promising and is still the best approach I can think of.

**Comment by AnizR** [comment]
This wasn't a good approach.
I have created another PR #716 that seems more promising.

**Pull request closed without merging**